pkg/storage/allocator_scorer.go

```diff
@@ -519,7 +529,7 @@ func rebalanceCandidates(
 					s, rangeInfo, balanceScore, sl)
 				continue
 			}
-			diversityScore := rebalanceToDiversityScore(s, existingNodeLocalities)
+			diversityScore := diversityRebalanceScore(existingCandidates[len(existingCandidates)-1].store, s, existingNodeLocalities)
```

Comment:
```
It seems likely that there are subtle issues here where the addition of the new candidate to the range could change which existing candidate would be chosen for removal.
```
Sorry, I couldn't quite understand what do you mean here. Yes, I'm just wondering the case if we will not remove the worst store from the existing replicas. We could make sure we'll not remove the replica we just rebalance to.

Reply:
Here's one of many potential examples. I'm sure there are more complicated ones, this is just the first that came to mind:

existing replica localities:
1. a=1,b=1,c=1
2. a=1,b=2,c=1
3. a=2,b=1,c=1

localities of possible rebalance candidates:
4. a=1,b=1,c=1
5. a=1,b=1,c=2

Since replicas 1 and 2 have equal diversity scores, we'll randomly choose one of them to consider the worst (ignoring which one of their stores is more loaded, which isn't relevant to this example). If we choose to remove replica 2, then we'll end up picking replica 5 to replace it. However, notice that this is actually hurting the diversity of the range. It'd be much better to remove replica 1 and replace it with either of 4 or 5.